### ISSUE TITLE
Weights File not saving

### ISSUE THREAD

**Issue by mad0511** [AlexeyAB/darknet#3313]
does the command automatically saves .weights file, if not how to save .weights file?

Mine's not saving .weights file 
![issue](https://user-images.githubusercontent.com/51075491/58862074-7fad6b00-86cd-11e9-8de8-ab29abef725c.png)

**Comment by maharleeeka** [comment]
what's inside your `obj.data`?

**Comment by mad0511** [comment]
> what's inside your `obj.data`?

This is my obj.data
![is](https://user-images.githubusercontent.com/51075491/58870056-0ae22d00-86dd-11e9-95d9-ab126f0941e0.png)

This is my obj.data file
![isss](https://user-images.githubusercontent.com/51075491/58870080-19304900-86dd-11e9-90f6-2373a1dd3256.png)

This is my obj.names file
![iss](https://user-images.githubusercontent.com/51075491/58870098-23524780-86dd-11e9-8023-f5794b1faeb1.png)

This is my train.txt file
![issss](https://user-images.githubusercontent.com/51075491/58870131-32d19080-86dd-11e9-9e62-bb4c09003c27.png)

**Comment by maharleeeka** [comment]
so there's nothing inside backup folder? 
everything looks fine to though, my weight files just auto saved inside backup folder

**Comment by mad0511** [comment]
No I didn't get any weights file in backup folder and also the training for 1600 images took only 20 seconds or so...my main doubt comes in this part whether it was trained or not? 

So kind of you to help

**Comment by kevinrev26** [comment]
Can you show the cfg file. It seems that is not configure properly.

**Comment by mad0511** [comment]
[net]
# Testing
# batch=1
# subdivisions=1
# Training
batch=64
subdivisions=16
width=608
height=608
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 500200
policy=steps
steps=45000,55000
scales=.1,.1

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

# Downsample

[convolutional]
batch_normalize=1
filters=64
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=32
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=128
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=64
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=256
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=512
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear


[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

# Downsample

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=2
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

[shortcut]
from=-3
activation=linear

######################

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
batch_normalize=1
filters=512
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=30
activation=linear


[yolo]
mask = 6,7,8
anchors = 11, 20,  18, 26,  15, 55,  28, 32,  35, 53,  52, 58,  52, 93,  82,101, 137,165
classes=5
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1


[route]
layers = -4

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 61



[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
batch_normalize=1
filters=256
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=512
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=30
activation=linear


[yolo]
mask = 3,4,5
anchors = 11, 20,  18, 26,  15, 55,  28, 32,  35, 53,  52, 58,  52, 93,  82,101, 137,165
classes=5
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1



[route]
layers = -4

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 36



[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=30
activation=linear


[yolo]
mask = 0,1,2
anchors = 11, 20,  18, 26,  15, 55,  28, 32,  35, 53,  52, 58,  52, 93,  82,101, 137,165
classes=5
num=9
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

**Comment by mad0511** [comment]
> Can you show the cfg file. It seems that is not configure properly.

For uploading in this section I used to save the cfg file in txt format...These are the parameters that I used for training my images with 5 classes..but the average IOU is displaying -nan values.


[yolo-obj.txt](https://github.com/AlexeyAB/darknet/files/3290746/yolo-obj.txt)

**Comment by dis-is-pj** [comment]
I got the same problem. Has anyone found the solution yet.

**Comment by andreasmarxer** [comment]
I also have the same problem. Trained about 8 hours on a cluster. Something doesn't work properly I guess.. I attached the log file, does someone have an idea?

[log.txt](https://github.com/AlexeyAB/darknet/files/3721617/log.txt)

**Comment by AlexeyAB** [comment]
@andreasmarxer Compile Darknet with GPU=1 CUDNN=1 in the Makefile

**Comment by andreasmarxer** [comment]
Thanks, man for the fast reply @AlexeyAB!

1. Unfortunately, I get the following error when setting GPU=1 and CUDNN=1 when compiling.:
![GPU=1_CompileError](https://user-images.githubusercontent.com/45263298/66714920-974bdd00-edac-11e9-93c0-609e75ddfd62.png)

(Previously I loaded the following modules specified in the ReadMe, as can be seen in the next picture)

2. When I also compile on OpenCV, due to the fact that I would like to see the actual training process on the cluster via the network, another error occurs:
![OpenCV=1_CompileError](https://user-images.githubusercontent.com/45263298/66714923-9d41be00-edac-11e9-9f91-f59606e552e4.png)

**Comment by AlexeyAB** [comment]
1. Compile darknet by using 
```
cmake .
make
```

2. install OpenCV

**Comment by andreasmarxer** [comment]
1. When doing `cmake . ` I get an Configuration incomplete, errors occured! output in the terminal.
The following log files are:
  [CMakeError.log](https://github.com/AlexeyAB/darknet/files/3721693/CMakeError.log)
  [CMakeOutput.log](https://github.com/AlexeyAB/darknet/files/3721694/CMakeOutput.log)

2. OpenCV is preinstalled on the cluster and I did load it, as seen in the 2nd picture.

**Comment by AlexeyAB** [comment]
@cenit Hi, do you know what is the reason?

**Comment by andreasmarxer** [comment]
> 1. When doing `cmake . ` I get an Configuration incomplete, errors occured! output in the terminal.
>    The following log files are:
>    [CMakeError.log](https://github.com/AlexeyAB/darknet/files/3721693/CMakeError.log)
>    [CMakeOutput.log](https://github.com/AlexeyAB/darknet/files/3721694/CMakeOutput.log)
> 
> The terminal output further says that Cmake needs to be 3.9+ and I have 3.8:
> ![3_cmake Error_CUDAsupport](https://user-images.githubusercontent.com/45263298/66715637-dd616a80-edc5-11e9-8e49-4b68f2229c66.png)
> 
> I updated the Cmake version, but the 2nd error about the CUDA version still exists:
> ![4_CUDA_VersionError](https://user-images.githubusercontent.com/45263298/66715642-e7836900-edc5-11e9-930b-7c39bcceab17.png)
> It says that I need Cuda version 8.0 exactly, why that? In your specifications, it's required 10.0. Furthermore, the error stays when I load Cuda version 8.0.61.
> 
> 1. OpenCV is preinstalled on the cluster and I did load it, as seen in the 2nd picture.

Actually, what's the difference between `cmake . && make` and only `make`?

I found out that the error also exists when I go back to the original Makefile with GPU=0 and CUDNN=0, but only when using the command `cmake .`. If I only run `make` it works for the original makefile with the following terminal output:
![6_GPU=0_CUDNN=0_make](https://user-images.githubusercontent.com/45263298/66717228-dabb4100-edd6-11e9-8355-ea6da8a9ad57.png)

Is this okay, or is also there an error?

**Comment by andreasmarxer** [comment]
In the meantime I got further experienting with different versions.
Actually I'm using CUDA 9.0.176, openCV 3.4.3, cudnn 7.0 and cmake 3.14.4.

The `cmake .` works without problems with this specifications.
In the `make` commant at 98% (when I already had the feeling I got it) the following error occurs:
![7_cmakeWorks_makeError98%](https://user-images.githubusercontent.com/45263298/66718260-1a3b5a80-ede2-11e9-8918-03ebd9701e0e.png)

Would be pleased getting further suggestions for solving this issue :) @AlexeyAB @cenit

**Comment by cenit** [comment]
Which cmake version are you using now after the update?

Also, please, *never* build with CMake **in-tree**. The best practice is always to build out of tree. So create another folder, wherever you want, then from there launch 
`cmake /path/to/darknet` 
`cmake --build . --target install`

Usually this build folder is just a subfolder of the project, and so the first command is just `cmake ..`

Anyway, did you just try to launch `./build.sh`? It should be the best and at the same time the easiest way

**Comment by cenit** [comment]
If you can’t build, it would be interesting to know the machine setup. Please let me know as many details as possible. It looks like an hpc cluster, it might be helpful to understand which one is that

**Comment by andreasmarxer** [comment]
Actually I'm using  CMake 3.14.4.

Okay, thanks for the hint to not build in-tree. I deleted the Microsoft Windows stuff in the build (darknet/build) folder and tried to build there with: 
`cmake ..`
This still writes me the build files into the folder _/cluster/home/marxera/darknet_ and not into the subfolder where I did run the command. 

Do I need to specify a path for the second command, because when I use it as stated above I get the **Error: Could not load cache.**

I also tried to use the build.sh command, but..
a)  when I use it in the darknet folder I get:
Error: could not load cache
b) when I use it in the subfolder (also as `./build.sh ..`) it says: No such file or directory.

The cluster is a HPC cluster, available for shareholders at my university ETH, named Leonhard Cluster. (https://scicomp.ethz.ch/wiki/Leonhard)
What detailed information than the versions mentioned above do you need?

I think I will also try to clean the repo and try again.

**Comment by andreasmarxer** [comment]
I have cleaned the repo now.
How would you suggest to proceed with the cleaned repository?

1. new folder _darknet/compile_
2. `cmake `.. in this folder
3. copy something into this folder for `cmake --build . --target install` or how to use this command?

**OR alternatively** only use the command `./build.sh`?
But don't I need to copy some files into the compile folder (because `./build.sh ..` didn't work) or is this command meant to be run in the main folder darknet?

**Comment by cenit** [comment]
Just running ./build.sh from the main folder is ok. The scripts does the magic. I am going to inspect your error and check the setup of the HPC cluster

**Comment by cenit** [comment]
inside the build.sh script, Below the third line, `force_cpp_build=false`, please add this line

```
additional_defines="-DBUILD_SHARED_LIBS=OFF"
```

and rebuild

**Comment by andreasmarxer** [comment]
It seemed it **finally worked** when adding the specified line into the build script!
![worked!!!](https://user-images.githubusercontent.com/45263298/66733409-c53e2980-ee5f-11e9-8762-0fef0937cb83.png)

Okay, that kind of makes sense.
Thank you very much for your help, that saved my day!
I'm now trying the training on my own dataset again on GPU and hope it's proceeding faster :)
